lib/tlog/tap.c

```diff
@@ -32,6 +32,117 @@
 #include <time.h>
 #include <semaphore.h>
 #include <assert.h>
+#include <limits.h>
+#include <ctype.h>
+#include <string.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+/**
+ * Add an utmp entry for this session.
+ *
+ * @param tap   tlog_tap structure address.
+ * @param path  The filename of the slave pty.
+ *
+ * @return zero, or -1 in case of error with errno set.
+ */
+static int
+tlog_tap_add_utmp_entry(struct tlog_tap *tap, char *path)
+{
+    int ofs;
+    struct timeval tv;
+    char *ses_line;
+    struct utmpx ut;
+    struct utmpx *ut_ptr;
+
+    if ((ses_line = ttyname(tap->tty_fd)) == NULL)
+        return -1;
+
+    if (gettimeofday(&tv, NULL) < 0)
+        return -1;
+
+    if ((tap->ut = calloc(sizeof(*tap->ut), 1)) == NULL)
+        return -1;
+
+    /* Set basic utmp entry data */
+    strncpy(ut.ut_user, getpwuid(getuid())->pw_name, sizeof(ut.ut_user));
```

Comment:
~~~
In file included from /usr/include/string.h:494,                                                                                                                                                                                              
                 from tap.c:37:                                                                                                                                                                                                               
In function 'strncpy',                                                                                                                                                                                                                        
    inlined from 'tlog_tap_add_utmp_entry' at tap.c:68:5,                                                                                                                                                                                     
    inlined from 'tlog_tap_setup' at tap.c:336:17:                                                                                                                                                                                            
/usr/include/bits/string_fortified.h:106:10: warning: '__builtin_strncpy' specified bound 32 equals destination size [-Wstringop-truncation]                                                                                                  
  106 |   return __builtin___strncpy_chk (__dest, __src, __len, __bos (__dest));                                                                                                                                                              
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~    
~~~